master/security/kubernetes-default-deny.md

```diff
@@ -0,0 +1,105 @@
+---
+title: Enable default deny for Kubernetes pods
+---
+
+### Big picture
+
+Change the default for Kubernetes pods from allowing all traffic, to denying all traffic using Kubernetes or Calico network policy.  
+
+### Value
+
+A **default deny** network policy provides an enhanced security posture -- so workloads without policy (or incorrect policy) are not allowed traffic until appropriate network policy is defined.
+
+### Features
+
+This how-to guide uses the following Calico features:
+- **NetworkPolicy** 
+- **GlobalNetworkPolicy**
+
+### Concepts
+
+#### Default deny/allow behavior
+
+**Default allow** means all traffic is allowed by default, unless otherwise specified. **Default deny** means all traffic is denied by default, unless explicitly allowed. **Kubernetes pods are default allow**, unless network policy is defined to specify otherwise.
+
+For compatibility with Kubernetes, **Calico network policy** enforcement follows the standard convention for Kubernetes pods:
+- If no network policies apply to a pod, then all traffic to/from that pod is allowed.
+- If one or more network policies apply to a pod containing ingress rules, then only the ingress traffic specifically allowed by those policies is allowed, and all other traffic to/from the pod is denied.
+- If one or more network policies apply to a pod containing egress rules, then only the egress traffic specifically allowed by those policies is allowed, and all other traffic to/from the pod is denied.
+
+For other endpoint types (VMs, host interfaces), **Calico global network policy is default deny**. Only traffic specifically allowed by network policy is allowed, even if no network policies apply to the endpoint.
+
+#### Secure resources everywhere
+
+We recommend creating default deny policy using **Calico global network policy**. It applies to both workloads (VMs and containers) and hosts (computers that run the hypervisor for VMs, or container runtime for containers). Using a Calico global network policy supports a conservative security stance for protecting resources.
+
+#### Best practice: implicit default deny policy
+
+We recommend that you create a "plain vanilla" default deny Calico network policy that does not include other deny rules. Why? The absence of other deny rules ensures that the default deny policy is implicitly "first/always". 
```

Comment:
I think we should probably reword or remove this "best practice" section. If we keep it, it should be focused on saying that a default deny itself is a best-practice since it ensures that unwanted traffic is denied by default.

I don't think we should spend any time on "explicit" denies here - those are actually used for a different use-case than "default deny" so the comparison is a bit out of place here. I think it also simplifies things since we don't need to talk about ordering here as well. 

We should call out (probably not here) the mechanism by which implicit deny rules work - specifically, its a combination of:
- Any pod selected by a policy gets its default changed from allow to deny.
- The implicit default-deny policy selects all pods, but does not allow any traffic via rules. 

> The absence of other deny rules ensures that the default deny policy is implicitly "first/always"

Also, the implicit deny always occurs _last_ - that is, if any other policy allows the traffic, then the deny won't come into affect. The deny is only executed _after_ all other policies have been evaluated.

Reply:
Fixed. Lots of good learning here. Thanks!